cmd/par/main.go

```diff
@@ -339,35 +341,49 @@ const (
 	eMemoryError                 = 8
 )
 
-func processVerifyOrRepairError(needsRepair bool, err error) {
+func processVerifyOrRepairError(needsRepair bool, err error) int {
 	// Match exit codes to par2cmdline.
 	if err != nil {
 		switch err.(type) {
 		case rsec16.NotEnoughParityShardsError:
 			fmt.Fprintf(os.Stderr, "Repair necessary but not possible.\n")
-			os.Exit(eRepairNotPossible)
+			return eRepairNotPossible
 		default:
 			fmt.Fprintf(os.Stderr, "Error encountered: %s\n", err)
-			os.Exit(eLogicError)
+			return eLogicError
 		}
 	}
 	if needsRepair {
 		fmt.Fprintf(os.Stderr, "Repair necessary and possible.\n")
-		os.Exit(eRepairPossible)
+		return eRepairPossible
 	}
-	os.Exit(eSuccess)
+	return eSuccess
 }
 
-func main() {
-	name := filepath.Base(os.Args[0])
+//export par_C
+func par_C(argc_ C.int, argv_ **C.char) int {
+	length := int(argc_)
+	cStrings := (*[1 << 28]*C.char)(unsafe.Pointer(argv_))[:length:length]
+
+	args_ := make([]string, length+1)
+	args_[0] = "libgopar"
+	for i, cString := range cStrings {
+		args_[i+1] = C.GoString(cString)
+	}
+	return par(args_)
```

Comment:
I looked at this, but "Recover is only useful inside deferred functions." threw me off. Should the whole function then go into a deferred function?

Reply:
No, only the recover needs to go in the deferred function. Basically, a deferred function is executed if the function exits either normally or via a panic (while the stack is unwinding). Then the `recover` function is how you check for that.

So in this case, it would look something like:

```
func par_c(...) {
    defer func() {
        if r := recover(); r != nil {
            // libgopar only panics with strings currently
            fmt.Println("libgopar panicked! ", r)
        }
    }()
  // rest of function
}